src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -108,6 +109,10 @@ public void Start<TContext>(IHttpApplication<TContext> application)
                         atLeastOneListener = true;
                         _disposables.Push(engine.CreateServer(
                             parsedAddress));
+
+                        // If requested port was "0", replace with assigned dynamic port.
+                        _serverAddresses.Addresses.Remove(address);
+                        _serverAddresses.Addresses.Add(parsedAddress.ToString());
```

Comment:
Have you tested with

* hostname
* IPv4 address
* IPv6 address

Wanna make sure this covers all bases (which it should, but need to see it working).